cpu/cc2538/mtarch.c

```diff
@@ -0,0 +1,164 @@
+void
+mtarch_start(struct mtarch_thread *thread,
+             void (*function)(void *data),
+             void *data)
+{
+  /* Calculate SP */
+  mtarch_stack_t *sp = (mtarch_stack_t *)&thread->stack[MTARCH_STACKSIZE - MTARCH_REGISTER_COUNT];
+  thread->sp = sp;
+
+  /* Zero stack */
+  memset(thread->stack, '\0', sizeof(void *) * MTARCH_STACKSIZE);
+
+  /* Fill stack */
+  sp->r0 = data;
+  sp->lr = mt_exit;
+  sp->pc = function;
+  sp->psr = (void *)0x01000000;
+}
+
+void
+mtarch_yield(void)
+{
+  /* Trigger PendSV the handler will switch to main */
+  REG(SCB_INTCTRL) = SCB_INTCTRL_PEND_SV;
+}
+
+void
+mtarch_exec(struct mtarch_thread *thread)
+{
+  /* Point to the next thread */
+  next_thread = thread;
+
+  /* Trigger PendSV, it will do the switch */
+  REG(SCB_INTCTRL) = SCB_INTCTRL_PEND_SV;
+}
+
+void
+mtarch_pstop(void)
+{
+}
+
+void
+mtarch_stop(struct mtarch_thread *thread)
+{
+}
+
+/**
+ * The schedule switches back and forth between
+ * main stack and thread stack(s). NULL pointers in
+ * next_thread and current_thread are used to indicate
+ * main stack.
+ *
+ * \param sp current stack pointer
+ * \return next stack pointer
+ */
+void *
+schedule(void *sp)
+{
+  if(current_thread != NULL) {
+    current_thread->sp = sp;
+  } else {
+    msp = sp;
+  }
+
+  /* Switch to next thread */
+  current_thread = next_thread;
+  next_thread = NULL;
+
+  if(current_thread) {
+    sp = current_thread->sp;
+  } else {
+    sp = msp;
+  }
+  return sp;
+}
+
+/* Disable prologue and epilogue of ISR */
+void pend_sv_isr(void)  __attribute__ ((isr, naked));
+
+/**
+ * The PendSV interupt service routine backs up current
+ * context, calls the scheduler and restores the newly
+ * selected context.
+ */
+void
+pend_sv_isr(void)
+{
+  /* Backup current thread context */
+  asm volatile ("cpsid i             \n"           /* Disable interrupts */
+                "dsb                 \n"           /* Data synchronization barrier */
+                "isb                 \n"           /* Instruction synchronization barrier */
+                "mrs r0, msp         \n"           /* Read SP to r0 */
+                "stmdb r0!, {r4-r11} \n"           /* Push registers onto stack */
+                "msr msp, r0");                    /* Update stack pointer */
+
+  /* Call sheduler Note: r0 is used as argument as well as result value */
+  asm volatile ("mov r4, lr\n"                     /* backup lr into callee saved register */
+                "bl schedule\n"                    /* Call scheduler */
+                "mov lr, r4\n");                   /* restore lr from callee saved register */
+
+  /* Restore next thread context */
+  asm volatile ("ldmfd r0!, {r4-r11} \n"           /* Pop registers from stack */
+                "msr msp, r0         \n"           /* Update stack pointer */
+                "cpsie i");                        /* Re-Enable interrupts */
+
+  asm volatile ("bx lr\n");                        /* Restore state and return  */
+}
```

Comment:
As you mentioned, this implementation applies to all Cortex-M3 platforms, so it should be moved to `cpu/arm/cortex-m3/`.